### PR TITLE
[DRAFT] Hash account_id. Update StateRecord

### PR DESCRIPTION
**Pull request by evgenykuzyakov**

### CODE REVIEW

File: core/store/src/lib.rs [hunk 1]
```diff
@@ -330,11 +335,19 @@ pub fn get_code(
         .map(|opt| opt.map(|code| ContractCode::new(code.to_vec())))
 }
 
+pub fn get_account_id_by_hash(
+    state_update: &TrieUpdate,
+    account_id_hash: CryptoHash,
+) -> Result<Option<AccountId>, StorageError> {
+    get(state_update, KeyForAccountId::from_hash(account_id_hash))
```
Comment (frol) [inline core/store/src/lib.rs]: How do we handle hash collision?

File: core/primitives/src/utils.rs [hunk 2]
```diff
@@ -73,19 +75,45 @@ impl KeyForAccount {
...
 #[derive(derive_more::AsRef, derive_more::Into)]
 #[as_ref(forward)]
 pub struct KeyForAccessKey(Vec<u8>);
 
 impl KeyForAccessKey {
-    fn estimate_prefix_len(account_id: &AccountId) -> usize {
-        KeyForColumnAccountId::estimate_len(col::ACCESS_KEY, account_id) + col::ACCESS_KEY.len()
+    #[inline]
+    fn estimate_prefix_len() -> usize {
+        KeyForColumnAccountId::estimate_len(col::ACCESS_KEY) + col::ACCESS_KEY.len()
```
Comment (frol) [inline core/primitives/src/utils.rs]: If you want to get rid of all the separators, there is one left here (it seems that `col::ACCESS_KEY` used to be used before and after the account_id).

File: core/store/src/lib.rs (hunk 1 above)
Comment (evgenykuzyakov) [reply core/store/src/lib.rs]: It's sha256

File: core/primitives/src/utils.rs [hunk 3]
```diff
@@ -132,36 +157,31 @@ impl KeyForAccessKey {
...
-    pub fn new(account_id: &AccountId, data: &[u8]) -> Self {
-        let mut key = Self::get_prefix_with_capacity(&account_id, data.len());
-        key.0.extend(data);
-        debug_assert_eq!(key.0.len(), Self::estimate_len(&account_id, &data));
-        key
+    pub fn new(account_id: &AccountId, key: &[u8]) -> Self {
+        let mut bytes = Self::get_prefix_with_capacity(&account_id, key.len());
+        bytes.0.extend(key);
+        debug_assert_eq!(bytes.0.len(), Self::estimate_len(&key));
+        bytes
```
[the remark this answers is not in the document]
Comment (evgenykuzyakov) [reply core/primitives/src/utils.rs]: `data_key` works

File: core/primitives/src/utils.rs [hunk 4]
```diff
@@ -73,19 +75,45 @@ impl KeyForAccount {
...
+    pub fn estimate_len() -> usize {
+        KeyForColumnAccountId::estimate_len(col::ACCOUNT_ID)
+    }
+
+    pub fn new(account_id: &AccountId) -> Self {
+        let key = KeyForColumnAccountId::with_capacity(col::ACCOUNT_ID, account_id, 0);
+        Self(key.into())
+    }
+
+    pub fn from_hash(account_hash: CryptoHash) -> Self {
```
[the remark this answers is not in the document]
Comment (evgenykuzyakov) [reply core/primitives/src/utils.rs]: Makes sense. My logic was the other keys are probably shouldn't be created from `account_hash` but instead should always be created from `account_id`.

File: core/primitives/src/utils.rs (hunk 2 above)
Comment (evgenykuzyakov) [reply core/primitives/src/utils.rs]: Oh. I see. Good point.

File: core/store/src/lib.rs (hunk 1 above)
Comment (frol) [reply core/store/src/lib.rs]: I still feel uneasy relying on the fact that there are no known collisions so far, but ok.

Comment (frol) [reply core/store/src/lib.rs]: Do we really need a hash function with strong cryptography properties for the trie key? I see it being cost-ineffective, and we also implicitly rely on a fact that there are no collisions (which is not true, there is just a very small chance), but that can lead us into a corrupted state when two accounts collide by the hash.

/cc @nearmax

**Comment by frol** [comment]
FYI, #2253 is merged. I had decoupled all the storage keys usage, so this PR should go smoothly with the state changes API.

NOTE: There are no unit-tests for the changes API (that would require a huge work on mocking various subsystems), so please, run `pytest/tests/sanity/rpc_state_changes.py` test once you are done.

File: core/store/src/lib.rs (hunk 1 above)
Comment (evgenykuzyakov) [reply core/store/src/lib.rs]: I think we're safe from the perspective of `account IDs` requiring a certain format. Even if a hash collision being discovered in the future, the chance that it fits `account ID` format is even lower.
We'll be able to upgrade state if suddenly there are sha256 collision being disovered.

**Comment by evgenykuzyakov** [comment]
@frol Seems, with the latest RPC API changes it's impossible to get account_id from a raw_key.

There are a few options:
- For every account hash from raw_key I need to lookup account_id from the trie as well. That's not an easy problem, because the mapping itself hash -> account_id might be deleted from the trie.
- Switch TrieUpdate APIs from using `raw_key` to take Trait that implements `raw_key` method. Then we can get account_id from Key.

**Comment by evgenykuzyakov** [comment]
> Switch TrieUpdate APIs from using raw_key to take Trait that implements raw_key method. Then we can get account_id from Key.

Also can't do this, because when we remove keys we might not have the original key, only the raw_key

**Third option** Cache account_hash -> account_id within TrieUpdate for every update you can make

**Comment by evgenykuzyakov** [comment]
> Third option Cache account_hash -> account_id within TrieUpdate for every update you can make

Might not work and requires a lot of changes

**Fourth option**
Store raw changes with `hash(account_id)` keys and map them back when a request comes in.

**Comment by evgenykuzyakov** [comment]
Blocked by #2278

**Comment by evgenykuzyakov** [comment]
Going to close it for now. It needs a lot of merges and probably has to be done slightly different

**Comment by frol** [comment]
@evgenykuzyakov Since we close this, can we at least somehow improve this hack: https://github.com/nearprotocol/nearcore/pull/2253#discussion_r392026033, as I feel uneasy leaving it there (I am not even sure what meaning 0x02 byte has in Borsh, which I use as the indicator for the beginning of the borsh-serialized public key).

**Pull request closed without merging**